Pull in fairscale.nn.Pipe into PyTorch.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44090 Pull in fairscale.nn.Pipe into PyTorch.**

This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

pritamdamania committed:
Pull in fairscale.nn.Pipe into PyTorch.

This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://fb.quip.com/txurAV3zIFox#RPZACAfAKMq. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://fb.quip.com/txurAV3zIFox#RPZACAfAKMq. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

msbaines reviewed (commented):
lgtm

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

pritamdamania87 commented on LICENSE:
```diff
@@ -16,23 +16,26 @@ Copyright (c) 2016-present, Facebook Inc. All rights reserved.
 All rights reserved.
- 
+
+All contributions by Kakao Brain:
```
license

pritamdamania87 commented on NOTICE:
```diff
@@ -22,6 +22,9 @@ All contributions by Yangqing Jia:
 Copyright (c) 2015 Yangqing Jia
 All rights reserved.
 
+All contributions by Kakao Brain:
```
notice

pritamdamania87 commented on test/distributed/_pipeline/sync/conftest.py:
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
```
license header

rohan-varma commented:
CI issue for Mac/Windows looks legit:
```
Oct 07 05:00:59 Traceback (most recent call last):
Oct 07 05:00:59   File "test/run_test.py", line 830, in <module>
Oct 07 05:00:59     main()
Oct 07 05:00:59   File "test/run_test.py", line 813, in main
Oct 07 05:00:59     raise RuntimeError(err_message)
Oct 07 05:00:59 RuntimeError: distributed/_pipeline/sync/skip/test_api failed!
```

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]

mrshenli commented on NOTICE:
```diff
@@ -22,6 +22,9 @@ All contributions by Yangqing Jia:
 All rights reserved.
 
+All contributions by Kakao Brain:
+Copyright 2019-2020 Kakao Brain
```
I assume the legal team has reviewed these license updates and they are OK with it?

pritamdamania87 replied:
Yes, this was added as per their guidance and I believe they mentioned they have a job that verifies licensing changes once things land. So they might reach out later if needed.

mrshenli commented on test/run_test.py:
```diff
@@ -95,6 +96,54 @@
     'test_fx_experimental',
     'test_functional_autograd_benchmark',
     'test_package',
+    'distributed/_pipeline/sync/skip/test_api',
+    'distributed/_pipeline/sync/skip/test_gpipe',
+    'distributed/_pipeline/sync/skip/test_inspect_skip_layout',
+    'distributed/_pipeline/sync/skip/test_leak',
+    'distributed/_pipeline/sync/skip/test_portal',
+    'distributed/_pipeline/sync/skip/test_stash_pop',
+    'distributed/_pipeline/sync/skip/test_tracker',
+    'distributed/_pipeline/sync/skip/test_verify_skippables',
+    'distributed/_pipeline/sync/test_balance',
+    'distributed/_pipeline/sync/test_bugs',
+    'distributed/_pipeline/sync/test_checkpoint',
+    'distributed/_pipeline/sync/test_copy',
+    'distributed/_pipeline/sync/test_deferred_batch_norm',
+    'distributed/_pipeline/sync/test_dependency',
+    'distributed/_pipeline/sync/test_inplace',
+    'distributed/_pipeline/sync/test_microbatch',
+    'distributed/_pipeline/sync/test_phony',
+    'distributed/_pipeline/sync/test_pipe',
+    'distributed/_pipeline/sync/test_pipeline',
+    'distributed/_pipeline/sync/test_stream',
+    'distributed/_pipeline/sync/test_transparency',
+    'distributed/_pipeline/sync/test_worker',
```
how much time does this add to general CI tests? If it is a lot, shall we consider adding it to slow tests for now to minimize the impact on other PRs?

pritamdamania87 replied:
Most of the tests didn't take too long, but I still added all of them to SLOW_TESTS to be safe.

mrshenli reviewed (approved):
Folder structure looks OK to me. 

Since we imported the code as-is from torch-gpipe, the current copyright makes sense. In the future, if we modified say 100% of one file, do we need to update the copyright in that file and the corresponding folder accordingly?

pritamdamania committed:
Update on "Pull in fairscale.nn.Pipe into PyTorch."


This is an initial commit pulling in the torchgpipe fork at
https://github.com/facebookresearch/fairscale.

The purpose of this commit is to just pull in the code and ensure all tests and
builds work fine. We will slowly modify this to match our intended API
mentioned in https://github.com/pytorch/pytorch/issues/44827. Follow up PRs would
address further changes needed on top of the initial commit..

We're pulling the code into the `torch.distributed._pipeline.sync` package. The
package is private on purpose since there is a lot of work (ex: docs, API
changes etc.) that needs to go in before we can actually officially support
this.

Differential Revision: [D23493316](https://our.internmc.facebook.com/intern/diff/D23493316/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23493316/)!

[ghstack-poisoned]
